optuna/storages/base.py

```diff
@@ -24,6 +24,69 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
 
     Storage classes abstract a backend database and provide library internal interfaces to
     read/write history of studies and trials.
+
+    **Thread safety**
+
+    Storage classes might be shared from multiple threads, and thus storage classes
+    must be thread-safe.
+    As one of the requirements of the thread-safety, storage classes must guarantee
+    that the returned values, such as `FrozenTrial`s will not be directly modified
+    by storage class.
+    However, storage class can assume that return values are never modified by users.
+    When users modify return values of storage classes, it might break the internal states
+    of storage classes, which will result in undefined behaviors.
+
+    **Ownership of RUNNING trials**
+
+    Trials in finished states are not allowed to be modified.
+    Trials in the WAITING state are not allowed to be modified except for the `state` field.
+    Storage classes can assume that each RUNNING trial is modified from only one process.
+    When users modify a RUNNING trial from multiple processes, it might lead to
+    an inconsistent internal state, which will result in undefined behaviors.
+    To use optuna with MPI or in other multi-process programs, users must make sure
+    that the optuna interface is accessed from only one of the processes.
+    Storage classes are not designed to provide inter-process communication functionalities.
+
+    **Consistency models**
+
+    Storage classes must support monotonic-reads consistency model, that is, if a
+    process reads a data `X`, any successive reads on data `X` does not return
+    older values.
+    They must support read-your-writes, that is, if a process writes to data `X`,
+    any successive reads on data `X` from the same process must read the written
+    value or one of more recent values.
+
+    **Stronger consistency requirements for special data**
+
+    TODO(ytsmiling) Add load method to storage class implementations.
```

Comment:
[NOTE] I'd like to discuss the name when this method is implemented (`load` sounds too general to me).

Reply:
I agree with this comment. I added a TODO comment in #1170.